### PR TITLE
BUG: read_csv used in file like object RawIOBase is not recognize encoding option

### LINKED ISSUE #31575: pandas 1.0.0  read_csv() is broken use `open( buffering=0)` option.
#### Code Sample

```python
import os
import pandas
import tempfile

fname = ""
with tempfile.NamedTemporaryFile(delete=False, mode="w+", encoding="shift-jis") as f:
        f.write("てすと\nbar")
        fname = f.name
print(fname)

try:
        with open(fname,mode="r", encoding="shift-jis") as f:
                result = pandas.read_csv(f)
                print("read shift-jis")
                print(result)

        with open(fname,mode="r", encoding="shift-jis") as f:
                result = pandas.read_csv(f,encoding="utf-8")
                print("open shift-jis file and read_csv with encoding: utf-8")
                print(result)

        with open(fname,mode="rb") as f:
                result = pandas.read_csv(f,encoding="shift-jis")
                print("open binary with buffered and read_csv with encoding: shift-jis")
                print(result)

        with open(fname,mode="rb",buffering=0) as f:
                result = pandas.read_csv(f,encoding="shift-jis")
                print("open binary without burrered and read_csv with encoding: shift-jis")
                print(result)
except Exception as e:
        print(e)

os.unlink(fname)
```
#### Problem description

Pandas 1.0.0, this sample does not work. But pandas 0.25.3, this sample works fine.

Open file with buffering=0 option, f is `RawIOBase`. This case seems encoding option will be ignored.

https://github.com/pandas-dev/pandas/pull/30771/files#diff-0335ae9037e4eb4747749a9f94cffd32R641

https://github.com/pandas-dev/pandas/pull/30771/files#diff-777d7549579ddc0c6e67596ad87e0d27R1879

#### Expected Output
```
/tmp/tmpxxxxxxxxxxx
read shift-jis
   てすと
0  bar
open shift-jis file and read_csv with encoding: utf-8
   てすと
0  bar
open binary with buffered and read_csv with encoding: shift-jis
   てすと
0  bar
open binary without burrered and read_csv with encoding: shift-jis
   てすと
0  bar
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.4.0-18362-Microsoft
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : ja_JP.UTF-8
LOCALE           : ja_JP.UTF-8

pandas           : 1.0.0
numpy            : 1.17.2
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 41.4.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by paihu**
- [x] closes #31575
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -109,6 +109,7 @@ Datetimelike
 - :meth:`DatetimeArray.searchsorted`, :meth:`TimedeltaArray.searchsorted`, :meth:`PeriodArray.searchsorted` not recognizing non-pandas scalars and incorrectly raising ``ValueError`` instead of ``TypeError`` (:issue:`30950`)
 - Bug in :class:`Timestamp` where constructing :class:`Timestamp` with dateutil timezone less than 128 nanoseconds before daylight saving time switch from winter to summer would result in nonexistent time (:issue:`31043`)
 - Bug in :meth:`DataFrame.reindex` and :meth:`Series.reindex` when reindexing with a tz-aware index (:issue:`26683`)
+- Bug in :meth: `read_csv` used in file like object `RawIOBase` is not recognize `encoding` option (:issue:`31575`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v1.1.0.rst]: Does this render correctly? (see [building the documentation](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#building-the-documentation))

I think you might need 
```
:meth:`read_csv`
```
(without the blank space)

Comment (paihu) [reply doc/source/whatsnew/v1.1.0.rst]: Oh, sorry. And change log position is not correct.

**Comment by TomAugspurger** [comment]
Is this going in 1.1 or 1.0.1? @jorisvandenbossche added the issue to 1.0.1, and if it's fixing a regression it should go there.

If so, you'll need to move the release note @paihu.

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -178,6 +178,7 @@ I/O
 ^^^
 - Bug in :meth:`read_json` where integer overflow was occuring when json contains big number strings. (:issue:`30320`)
 - Bug in :meth:`DataFrame.to_json` was raising ``NotFoundError`` when ``path_or_buf`` was an S3 URI (:issue:`28375`)
+- Bug in :meth:`read_csv` used in file like object ``RawIOBase`` is not recognize ``encoding`` option (:issue:`31575`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: you can move to 1.0.1 (same section)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. pls move the note; ping on green.

**Comment by jreback** [comment]
@paihu make sure to merge master you have a conflict

File: doc/source/whatsnew/v1.1.0.rst (hunk 2 above)
Comment (paihu) [reply doc/source/whatsnew/v1.1.0.rst]: I moved it.

**Comment by paihu** [comment]
I think this issue is regression bug. so I insert `Fix regressions` section.

**Review by jorisvandenbossche — COMMENTED** [review]
Should we rather check for `isinstance(.., io.IOBase)` instead of checking both BufferedIOBase and RawIOBase? 

`IOBase` additionally is also the base class of TextIOBase

cc @gfyoung

**Comment by paihu** [comment]
We don't want to wrap text(`TextIOBase`). We need to wrap only binary(`BufferedIOBase` and `RawIOBase`).

**Comment by jorisvandenbossche** [comment]
Because in that case (eg something like a StringIO object), it is assumed to be already encoded UTF string, so encoding shouldn't be applied then?

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @paihu!

**Comment by gfyoung** [comment]
Thanks @paihu!

**Pull request merged**